File: src/app/weapons/projectile/projectile-sfx.lua
local ____exports = {}
____exports.ProjectileSFX = {}
local ProjectileSFX = ____exports.ProjectileSFX
ProjectileSFX.name = "ProjectileSFX"
ProjectileSFX.__index = ProjectileSFX
ProjectileSFX.prototype = {}
ProjectileSFX.prototype.__index = ProjectileSFX.prototype
ProjectileSFX.prototype.constructor = ProjectileSFX
function ProjectileSFX.new(...)
    local self = setmetatable({}, ProjectileSFX.prototype)
    self:____constructor(...)
    return self
end
function ProjectileSFX.prototype.____constructor(self, sfx, startingLoc, offset, facing)
    self.offset = offset
    self.yaw = Atan2(facing.y, facing.x)
    self.pitch = Asin(facing.z)
    self.roll = 0
    self.sfx = AddSpecialEffect(sfx, startingLoc.x, startingLoc.y)
    BlzSetSpecialEffectZ(self.sfx, startingLoc.z)
    BlzSetSpecialEffectRoll(self.sfx, self.pitch)
    BlzSetSpecialEffectYaw(self.sfx, self.yaw)
end
function ProjectileSFX.prototype.updatePosition(self, currentPosition)
    BlzSetSpecialEffectX(self.sfx, currentPosition.x + self.offset.x)
    BlzSetSpecialEffectY(self.sfx, currentPosition.y + self.offset.y)
    BlzSetSpecialEffectZ(self.sfx, currentPosition.z + self.offset.z)
end
function ProjectileSFX.prototype.setScale(self, scale)
    BlzSetSpecialEffectScale(self.sfx, scale)
end
function ProjectileSFX.prototype.destroy(self)
    DestroyEffect(self.sfx)
end
return ____exports

File: src/app/weapons/projectile/projectile-sfx.ts
import { Vector3 } from "../../types/vector3";
import { getYawPitchRollFromVector } from "../../../lib/translators";

export class ProjectileSFX {
    private offset: Vector3;
    private sfx: effect;

    private pitch: number;
    private roll: number;
    private yaw: number;

    constructor(sfx: string, startingLoc: Vector3, offset: Vector3, facing: Vector3) {
        // Rotate by facing
        this.offset = offset;

        const facingData = getYawPitchRollFromVector(facing);
        this.yaw = facingData.yaw;
        this.pitch = facingData.pitch;
        this.roll = facingData.roll;

        this.sfx = AddSpecialEffect(sfx, startingLoc.x, startingLoc.y); 
        BlzSetSpecialEffectZ(this.sfx, startingLoc.z);
        BlzSetSpecialEffectPitch(this.sfx, this.pitch);
        BlzSetSpecialEffectYaw(this.sfx, this.yaw);
    }

    updatePosition(currentPosition: Vector3, velocityChange?: Vector3) {
        BlzSetSpecialEffectPosition(this.sfx,
            currentPosition.x + this.offset.x,
            currentPosition.y + this.offset.y,
            currentPosition.z + this.offset.z
        );

        if (velocityChange) {
            BlzSetSpecialEffectPitch(this.sfx, -1*Math.atan2(velocityChange.z, velocityChange.to2D().getLength()));
        }
    }

    setScale(scale: number) {
        BlzSetSpecialEffectScale(this.sfx, scale);
    }

    destroy() {
        DestroyEffect(this.sfx);
    }

    getEffect() {
        return this.sfx;
    }
}
